File: packages/shared/src/cognitive/position-event.ts
/**
 * PositionEvent — vista canónica de ExitReason (V1.44).
 * No sustituye ExitPlan. No es autoridad. No persiste.
 */

import type { ExitReasonV1 } from "./exit-plan.js";

export type PositionEventKindV1 =
  | "STOP"
  | "T1"
  | "T2"
  | "TRAIL"
  | "INVALIDATION"
  | "TIME"
  | "PORTFOLIO_RISK"
  | "MANUAL";

export type PositionEventV1 = {
  kind: PositionEventKindV1;
  reasonCode: ExitReasonV1;
  at: string;
};

const KIND_BY_REASON: Record<ExitReasonV1, PositionEventKindV1> = {
  STRUCTURAL_STOP: "STOP",
  TARGET_1: "T1",
  TARGET_2: "T2",
  TRAIL: "TRAIL",
  THESIS_INVALIDATION: "INVALIDATION",
  TIME_STOP: "TIME",
  PORTFOLIO_RISK: "PORTFOLIO_RISK",
  MANUAL: "MANUAL",
};

export function positionEventKindFromReason(
  reason: ExitReasonV1 | null | undefined,
): PositionEventKindV1 | null {
  if (!reason) return null;
  return KIND_BY_REASON[reason] ?? null;
}

export function buildPositionEvent(
  reason: ExitReasonV1 | null | undefined,
  at: string,
): PositionEventV1 | null {
  const kind = positionEventKindFromReason(reason);
  if (!kind || !reason) return null;
  const stamp = typeof at === "string" && at.trim() ? at : "";
  if (!stamp) return null;
  return { kind, reasonCode: reason, at: stamp };
}

export function isTargetEventKind(kind: PositionEventKindV1 | null): boolean {
  return kind === "T1" || kind === "T2" || kind === "TIME";
}

export function isImmediateRiskReason(
  reason: ExitReasonV1 | null | undefined,
): boolean {
  return (
    reason === "STRUCTURAL_STOP" ||
    reason === "THESIS_INVALIDATION" ||
    reason === "PORTFOLIO_RISK"
  );
}

File: packages/py/analytics/src/bolsa_analytics/cognitive/position_event.py
"""PositionEvent — vista canónica de ExitReason (V1.44). ≠ ExitPlan."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from bolsa_analytics.cognitive.exit_plan import ExitReason

PositionEventKind = Literal[
    "STOP",
    "T1",
    "T2",
    "TRAIL",
    "INVALIDATION",
    "TIME",
    "PORTFOLIO_RISK",
    "MANUAL",
]

_KIND_BY_REASON: dict[str, PositionEventKind] = {
    "STRUCTURAL_STOP": "STOP",
    "TARGET_1": "T1",
    "TARGET_2": "T2",
    "TRAIL": "TRAIL",
    "THESIS_INVALIDATION": "INVALIDATION",
    "TIME_STOP": "TIME",
    "PORTFOLIO_RISK": "PORTFOLIO_RISK",
    "MANUAL": "MANUAL",
}


@dataclass(frozen=True, slots=True)
class PositionEvent:
    kind: PositionEventKind
    reason_code: ExitReason
    at: str


def position_event_kind_from_reason(reason: str | None) -> PositionEventKind | None:
    if not reason:
        return None
    return _KIND_BY_REASON.get(reason)


def build_position_event(reason: str | None, at: str) -> PositionEvent | None:
    kind = position_event_kind_from_reason(reason)
    if kind is None or not reason or not isinstance(at, str) or not at.strip():
        return None
    return PositionEvent(kind=kind, reason_code=reason, at=at)  # type: ignore[arg-type]


def is_target_event_kind(kind: PositionEventKind | None) -> bool:
    return kind in ("T1", "T2", "TIME")


def is_immediate_risk_reason(reason: str | None) -> bool:
    return reason in (
        "STRUCTURAL_STOP",
        "THESIS_INVALIDATION",
        "PORTFOLIO_RISK",
    )
